### DISCUSSION TITLE
Add BaseDirectoey.root

### DISCUSSION THREAD

**Discussion by 781flyingdutchman** [781flyingdutchman/background_downloader#185] [Ideas]
Allows upload/download to any destination, with the risk related to app sandbox moving to the developer.

Not ideal, but there are so many questions about this that I think it's worth adding in V8

**Comment by FluffyDiscord** [comment]
I really need to be able to download and upload files directly to/from somewhere in user created folders on Android. There is already package, that can read/write anywhere that user permits using Android's SAF. Would you consider enabling/adding support to download anywhere? https://github.com/alexrintt/shared-storage / https://developer.android.com/training/data-storage/manage-all-files

Right now it's impossible to do so with background_download. We can only move files to some predefined shared folders. Uploading is in even worse state. We have to copy the file to the scoped storage in order to not lose it, since renaming/moving makes the file disappear in the span of the uploading time.

Maybe this could be worked around by adding save/read callbacks where we could handle the IO ourselves? Not sure if possible, or if that would be harder to implement.

**Comment by 781flyingdutchman** [comment]
Yes, I'm planning this for V8, which is what I'm worrying right now. The way this would work is:

1. Set `baseDirectory` to `BaseDirectory.root` (a new value in the `BaseDirectory` enum
2. The task's `directory` field then becomes the absolute path to a directory on the system, and `filename` completes that path to point to a specific file
3. Provided you have access to that path, you can download to or upload from that path

You should still use the other `BaseDirectory` values whenever possible to avoid issues with unstable application directories. If you use `.root` you're "on your own" w.r.t. stability and permissions.

**Comment by 781flyingdutchman** [comment]
@FluffyDiscord I've added this to branch `V8-dev`. Do you mind checking it out before this is released as V8?

From the readme: You can also pass an absolute path to the downloader by using `BaseDirectory.root` combined with the path in `directory`. This allows you to reach any file destination on your platform. However, be careful: the reason you should not normally do this (and use e.g. `BaseDirectory.applicationDocuments` instead) is that the location of the app's documents directory may change between application starts (on iOS, and on Android in some cases), and may therefore fail for downloads that complete while the app is suspended.  You should therefore never store permanently, or hard-code, an absolute path, unless you are absolutely sure that that path is 'stable'.

**Reply by FluffyDiscord** [reply]
I will test it later today.

**Reply by FluffyDiscord** [reply]
Works, but I think you should enable to have the leading `/` when `BaseDirectory.root` is being used, so we don't have to always strip the first slash (basically the check at `package:background_downloader/src/task.dart:296:7`)

**Comment by 781flyingdutchman** [comment]
Good suggestion, I'll change that

**Comment by 781flyingdutchman** [comment]
Included in V8, and the leading path separator is now always stripped from the `directory` field.
